new-packages/ts-project/src/state.ts

```diff
@@ -0,0 +1,203 @@
+            } else if (hasUnefinedValue(ts, prop.initializer)) {
+              result.data[key] = undefined;
+            } else {
+              ctx.errors.push({
+                type: 'state_property_unhandled',
+                node: prop,
+              });
+            }
+            break;
+          }
+          case 'type': {
+            const isStringWithValidValue =
+              ts.isStringLiteral(prop.initializer) &&
+              ['history', 'atomic', 'compound', 'parallel', 'final'].includes(
+                prop.initializer.text,
+              );
+
+            if (isStringWithValidValue) {
+              result.data[key] = prop.initializer
+                .text as ExtractorStateConfig['data']['type'];
+            } else if (hasUnefinedValue(ts, prop.initializer)) {
+              result.data[key] = undefined;
+            } else {
+              ctx.errors.push({
+                type: 'state_property_unhandled',
+                node: prop,
+              });
+            }
+            break;
+          }
+          // TODO: this property only has effect if type is set to history
+          case 'history': {
+            const isStringWithValidValue =
+              ts.isStringLiteral(prop.initializer) &&
+              ['shallow', 'deep'].includes(prop.initializer.text);
+
+            if (isStringWithValidValue) {
+              result.data[key] = prop.initializer
+                .text as ExtractorStateConfig['data']['history'];
+            } else if (hasUnefinedValue(ts, prop.initializer)) {
+              result.data[key] = undefined;
+            } else {
+              ctx.errors.push({
+                type: 'state_property_unhandled',
+                node: prop,
+              });
+            }
+            break;
+          }
+          default:
+        }
+      }
+      continue;
+    }
+
+    if (ts.isShorthandPropertyAssignment(prop)) {
+      ctx.errors.push({
+        type: 'state_property_unhandled',
+        node: prop,
+      });
+      continue;
+    }
+    if (ts.isSpreadAssignment(prop)) {
+      ctx.errors.push({
+        type: 'state_property_unhandled',
+        node: prop,
+      });
+      continue;
+    }
+    if (
+      ts.isMethodDeclaration(prop) ||
+      ts.isGetAccessorDeclaration(prop) ||
+      ts.isSetAccessorDeclaration(prop)
+    ) {
+      ctx.errors.push({
+        type: 'state_property_unhandled',
+        node: prop,
+      });
+      continue;
+    }
+
+    prop satisfies never;
+  }
+
+  return {
+    id: path.join('.'),
+    data: {
+      initial: result.data.initial ?? undefined,
+      type: result.data.type ?? undefined,
+      history: result.data.history ?? undefined,
+      metaEntries: result.data.metaEntries ?? [],
+      entry: result.data.entry ?? [],
+      exit: result.data.exit ?? [],
+      invoke: result.data.invoke ?? [],
+      tags: result.data.tags ?? [],
+      description: result.data.description ?? undefined,
```

Comment:
I'm not a fan of this `data` nesting 😅 This is how digraph keeps it though so we have to keep it here too :(